Document the option to have `Option` in queries

mlange-42 opened the pull request:
Not sure if the example is a bit too much? Maybe a one- or three-liner would also suffice? Alternatively, we could integrate it into the primary Query example.

adamreichold commented:
> Alternatively, we could integrate it into the primary Query example.

Yes, I would prefer that approach. The sentence on how this works should not be lost though. I think it would be best if we started the documentation of `struct Query` with a sentence explaining that queries are specified by composing the type of their results, i.e.
* `&C` for shared access
* `&mut C` for exclusive access
* `(P, Q)` to combine two queries into one
* `Option<Q>` for optional components
*  `With<Q, C>` to filter `Q` for presence of `C`
* `Without<Q,C>` to filter `Q` for absence of `C`

I would also like it if the example you unpack the `Option` to demonstrate that one can access the item of the inner query if present. Ideally, put that might be pushing it w.r.t. readability, we would put a tuple inside the like `Option<(&Foo, &mut Bar)>` to demonstrate that its argument is a query itself, not just a component.

mlange-42 commented:
Listing all the possibilities explicitly sounds good, as does integration into a single example. Tuples inside `Option` are another thing I was not aware of.

mlange-42 commented:
Thinking about options (of tuples) a bit, some questions came up to me... it is possible to have `Option<Foo>` as a component, right? Wouldn't that make `Option`s in Queries ambiguous? 

What if one adds components like `(i32, bool, Option<i32, bool>)`,  and uses `Query<(Option<i32, bool>, )>`?

adamreichold commented:
> Wouldn't that make Options in Queries ambiguous?

No, as the bare query types always contain references to types, not the bare types themselves.

> What if one adds components like (i32, bool, Option<(i32, bool)>), and uses Query<(Option<i32, bool>, )>?

Querying for `Option<(i32, bool)>` should not compile as there is not query impl for `i32` and `bool`. Querying for the mandatory components would be done using `(&i32, &bool)` and querying for the optional component would be `&Option<(i32, bool)>`. (And optionally querying for the optional component would be `Option<&Option<(i32, bool)>>`.

As a general remark, even if there was a problem, there wouldn't be in realistic examples as we would probably not store `Option<bool>` as a component but rather `SomeNewTypeWrapper(Option<bool>)` to illustrate what the `bool` is actually about. And instead of using `Option<SomeNewTypeWrapper>` we would have the component present or not.

mlange-42 commented:
Ah, right... I somehow missed the ref in the queries while thinking about the Options stuff, AFK. Of course, this resolves it.

mlange-42 commented:
Implemented the suggested list of possible query specifications. Also added a sentence on Entities as Components.

Could it be helpful to also point out how to store a Query in a System? I.e. the lifetime specifiers in 

```rust
query = Query<(&'static C, &'static mut D)>,
```

adamreichold commented:
> Could it be helpful to also point out how to store a Query in a System? I.e. the lifetime specifiers in

Certainly, I am just wondering where to put this? Integrate into the paragraph starting with `Queries are provided as stand-alone structs`? Adding to the documentation of `Query::new` to avoid overloading that part? But then, we would show using `Default::default` or even `#[derive(Default)]` for initialisation?

adamreichold commented:
Actually just putting that as a second example on the struct doc would probably be most natural and easy to find?

mlange-42 commented:
Added a second example demonstrating prepared queries.

adamreichold commented on src/query.rs:
```diff
@@ -22,17 +35,55 @@ use crate::{
+///     world.insert(entity, (0_i32, true));
+///
+///     let mut system = System::default();
+///     system.update(&world);
```
Maybe put this into some kind of loop to really reuse the system?

mlange-42 replied:
Yes, that makes sense.

adamreichold replied:
Will merge and amend the loop. (Just doing the tag and publish before calling it a day.)
